### ISSUE TITLE
Replacing Travis CI

### ISSUE THREAD

**Issue by bbatsov** [flycheck/flycheck#1957]
Travis CI has been discontinued and we need some alternative. It seems pretty much all PRs are block on this. That's the final build that ever ran:

![image](https://user-images.githubusercontent.com/103882/179763085-381fa563-ef0f-4e21-8bdb-3536cf878f71.png)

Should we just remove it for now while we're in the process of identifying a replacement?

**Comment by cpitclaudel** [comment]
Yes, let's remove it.  For now I've removed the branch protection.
FWIW, I think it's good to run the flycheck tests, but the tests that verify that we can parse the output of each and every checker are too much work (because they depend on the exact phrasing of each message.

**Comment by bbatsov** [comment]
Supposedly there's still some Travis CI that can be used with relatively few changes - https://docs.travis-ci.com/user/migrate/open-source-repository-migration Perhaps this should be explored at some point. I've noticed that one major issue with the tests is that we have to install a ton of tools to check their output. I assume those are the tests that you're referring to.

**Comment by jcs090218** [comment]
I would love to work on this if flycheck would like to use [Eask](https://github.com/emacs-eask/cli) for their CI tests. Eask is very similar to Cask, so it shouldn't be hard to do it.

BTW, Eask has built-in support to read files like Eask.24, Eask.25.1, etc. Multiple Eask-file co-exists is possible and should make CI tests a bit easier for this package.

**Comment by bbatsov** [comment]
These days there's also Eldev, which looks similar to Eask. Not sure what advantages Eask might have over Eldev.

**Comment by jcs090218** [comment]
I know Eldev, but I am not very familiar with it. Can you explain the advantages of using Eldev?

Eask is similar to Cask if we compare two files:

- Example Cask file: https://github.com/flycheck/flycheck/blob/master/Cask (from this repo)
- Example Eask file: https://github.com/emacs-lsp/lsp-mode/blob/master/Eask

The commands are similar as well, 

```bash
$ cask eval
$ cask install

$ eask eval
$ eask install
```

Eask is intended to build similar to Cask but with better cross-platform capability (using node), simpler wrapper, etc.

**Comment by bbatsov** [comment]
Well, I can't speak for Eask (having never used it), but compared to Cask I like that `eldev` supports things like linting and running tests with various tools out-of-the-box. It also have configuration that's written in Elisp, instead of some DSL, which for me is a feature. 

With Eldev the CI pipelines for my projects became super simple - e.g. https://github.com/clojure-emacs/cider/blob/master/.circleci/config.yml 

@doublep Can probably say a lot more on how Eldev compares to Cask (and potentially Eask).

**Comment by cpitclaudel** [comment]
What's wrong with Cask? (I'm not sure how cross-platform it is, but it's written in a mix of ELisp and Python, and at least on Ubuntu python3 comes out of the box and node needs to be installed, so I'm not sure node helps very much)

**Comment by jcs090218** [comment]
> Well, I can't speak for Eask (having never used it), but compared to Cask I like that eldev supports things like linting and running tests with various tools out-of-the-box.

#### Out-of-box linting via `eask lint`

```
Run linter

Usage: eask lint <command> [options..]

Commands:
  lint checkdoc [files..]     Run checkdoc
  lint declare [files..]      Run check-declare
  lint elint [files..]        Run elint
  lint elisp-lint [files..]   Run elisp-lint
  lint elsa [files..]         Run elsa
  lint indent [files..]       Run indent-lint
  lint keywords               Lint the package's keywords header
  lint package [files..]      Run package-lint
  lint regexps [files..]      Run relint   
...
```

#### Out-of-box test via `eask test`

```
Run test

Usage: eask test <command> [options..]

Commands:
  test activate [files..]     Activate package; use to test package activation
  test buttercup [files..]    Run buttercup tests
  test ert-runner [files..]   Run ert tests using ert-runner
  test ert [files..]          Run ert tests
...
```

> It also have configuration that's written in Elisp, instead of some DSL, which for me is a feature.

Eask uses elisp, DSL is optional. Eask-file is an elisp file. You can treat it as `init.el` file but for project only.

> With Eldev the CI pipelines for my projects became super simple - e.g. https://github.com/clojure-emacs/cider/blob/master/.circleci/config.yml

I don't think it's simple... but here are my usage for CI pipelines with Eask:

- https://github.com/emacs-grammarly/lsp-grammarly/blob/master/.github/workflows/test.yml
- https://github.com/emacs-lsp/lsp-mode/blob/master/.github/workflows/test.yml (for larger project, it's more complicated)

In conclusion, Eask is mixture of Cask and Eldev.

---

> What's wrong with Cask? (I'm not sure how cross-platform it is, but it's written in a mix of ELisp and Python, and at least on Ubuntu python3 comes out of the box and node needs to be installed, so I'm not sure node helps very much)

Python is only required for Windows, you don't need Python if you are on Linux or macOS. Cask works but it is basically unmaintained and so a "risk".

`@ubolonton` has a comment on reddit well explain the issue regarding Cask:

> For me, Cask's biggest problems are:
> 
> Bootstrapping (into Emacs Lisp) is unreliable, because it's based on Python and shell scripts. The situation is particularly bad on Windows.
> 
> * There is no concise way to define project-specific tasks.
> * Solutions based on Makefile solve the second issue, but is not ideal on Windows.
> 
> I ended up writing custom Bash/PowerShell scripts.

Source: https://www.reddit.com/r/emacs/comments/fmkk9s/comment/fl6v30u/?utm_source=reddit&utm_medium=web2x&context=3

Eask address all these problems.

Another major issue is Python is particularly bad on Windows (someone claims it's bad on Unix-like system as well, but Windows is the worst). I have worked with both environments (Python and Node), and node won the battle on cross-platform capability. I don't know "how well" is it, but I never had issue with node, so I chose it.

**Comment by cpitclaudel** [comment]
The Windows discussion is interesting (IME python works just fine there, and node + emacs on Windows is known to have [issues](https://github.com/nodejs/node/issues/4697)). But in any case our CI doesn't run on Windows, right?

**Comment by jcs090218** [comment]
> The Windows discussion is interesting (IME python works just fine there, and node + emacs on Windows is known to have https://github.com/nodejs/node/issues/4697). 

My experiences with Python on Windows is "runnable", but I won't consider "good". Python on Linux is "smooth and okay".

I have looked into https://github.com/nodejs/node/issues/4697. Eask only uses node for activation (shell layer), then it will eventually be handed to Emacs (executable), so I am not too worried about this I/O issue since we can always fall back to the Emacs layer and not node layer. And another way around, if something does not work in Emacs, we could let the node does the task.

> But in any case our CI doesn't run on Windows, right?

I would suggest adding some basic testing, like `packaging`, `installing`, `compile`, etc. So you know there won't be breakage bugs on MELPA and client ends. `lsp-mode` does something like that since most developers in the org are mostly Linux users.

**Comment by cpitclaudel** [comment]
Thanks!  I'll say one thing's certain for me: if you're willing to resurrect our CI, and the result seems reasonably maintainable (not a very high bar, just something that others can edit if needed), then I'll be very happy to take a dependency on Eask ^^

**Comment by bbatsov** [comment]
That's fair, although I think we should first just move what we currently have to GitHub Actions or CircleCI and take it from there. That's why I think the first decision to make is what to replace Travis with.

**Comment by jcs090218** [comment]
> Thanks! I'll say one thing's certain for me: if you're willing to resurrect our CI, and the result seems reasonably maintainable (not a very high bar, just something that others can edit if needed), then I'll be very happy to take a dependency on Eask ^^

My plan is to help maintain Flycheck, so if something breaks, then I will be there and responsible for the mission. But yes! I would like to keep maintenance low, so other maintainers/developers can still keep up with the project status! ;)

> That's fair, although I think we should first just move what we currently have to GitHub Actions or CircleCI and take it from there. That's why I think the first decision to make is what to replace Travis with.

I might go for GHA since most of my projects use it.

**Comment by Fuco1** [comment]
FWIW I love Eask and I'm migrating all my projects to it. @jcs090218 is super responsive and I've submitted tons of issues/feature requests and he dispatched them very fast.

The Eask UX is unparalleled in Emacs world, startup is very fast, the suggestions when something breaks are actually helpful, many things work out of the box (like 10+ linters, all the usual testing frameworks etc).  The other build tools are pretty much dead except eldev, which I haven't used much.

**Comment by cpitclaudel** [comment]
That's wonderful to hear! :)  Now I'm positively excited that we're migrating ^^ 
Thanks @jcs090218 for your excellent work and happy to have you onboard!

**Comment by casch-at** [comment]
Correct me if I'm wrong, but it seems like this issue can be closed.

**Comment by jcs090218** [comment]
> Correct me if I'm wrong, but it seems like this issue can be closed.

👍
